modules/saku-policy-store/src/saku/schemas.clj

```diff
@@ -1,53 +1,198 @@
-
-(def resource-statement
-  (m/schema [:map
-             [:statement/actions [:sequential :string]]
-             [:statement/effect effect]
-             [:statement/identities [:sequential :string]]]))
+(defn- -compile-rename-keys-transformer
+  [kmap]
+  {:leave (if (seq kmap)
+            #(rename-keys % kmap)
+            identity)})
 
-(def resource-policy
-  (m/schema [:map
-             [:policy/drn :string]
-             [:policy/statements [:sequential {:min 1} resource-statement]]]))
-
-(def resource-policies
-  (m/schema [:sequential resource-policy]))
-
-(def identity-policy
-  (m/schema [:map
-             [:policy/drn :string]
-             [:policy/statements [:sequential {:min 1} identity-statement]]]))
+(defn rename-keys-transformer
+  [{:keys [rename-key]}]
+  (mt/transformer
+    {:decoders
+     {:map
+      {:compile
+       (fn [schema _]
+         (-compile-rename-keys-transformer
+           (into {}
+             (keep (fn [[k-src val-schema]]
+                     (when-let [rename (get (m/properties val-schema) rename-key)]
+                       [k-src rename])))
+             (m/entries schema))))}}}))
 
-(def identity-policies
-  (m/schema [:sequential identity-policy]))
+(def PolicyType (m/schema [:enum "Resource" "Identity"]))
 
-(def evaluate-one-args
+(def EvaluateOneInput
   (m/schema [:map
              [:drn :string]
              [:actionId :string]
              [:identities [:sequential {:min 1} :string]]]))
 
-(def evaluate-many-args
+(def EvaluateManyInput
   (m/schema [:map
              [:drns [:sequential {:min 1} :string]]
              [:actionId :string]
              [:identities [:sequential {:min 1} :string]]]))
 
-(defn assert*
-  ([schema obj]
-   (when-not (m/validate schema obj)
-     (throw (ex-info "Invalid type" {:anomaly/category ::invalid-type
-                                     :schema schema
-                                     :obj obj
-                                     :cause (m/explain schema obj)})))))
+(def StatementInput
+  [:map
+   [:sid {:optional true
+          :output-k :statement/sid} string?]
+   [:actionIds {:output-k :statement/actions} [:sequential string?]]
+   [:effect {:output-k (fn [x] [:statement/effect [:effect (csk/->kebab-case-keyword x)]])} string?]])
+
+(def ResourceStatementInput
+  (mu/merge StatementInput
+    [:map
+     [:identities {:output-k :statement/identities} [:sequential string?]]]))
+
+(def IdentityStatementInput
+  (mu/merge StatementInput
+    [:map
+     [:resources {:output-k :statement/resources} [:sequential string?]]]))
+
+(defn unique-sids?
+  [statements]
+  ;; we ignore `nil` sids since those do not need to be unique
+  (let [sids (keep :sid statements)]
+    (= (distinct sids) sids)))
+
+(def UniqueSids
+  [:fn {:error/fn (fn [{:keys [value]} _]
+                    (str "policy must have unique statement ids, got duplicate `sid`s: " (str/join ", " (keep :sid value))))}
+   unique-sids?])
```

Comment:
This behavior is actually missing from the recent https://github.com/luchiniatwork/saku/pull/15. If you upsert a policy, you could end up with two statements having the same `sid`. This is almost certainly not what you want. As such, we prevent it at the input level. A test for this is also included.